### ISSUE TITLE
Support ligatures in terminal

### ISSUE THREAD

**Issue by Tyriar** [microsoft/vscode#34103]
Regresses with performance improvements in terminal's renderer https://github.com/Microsoft/vscode/pull/33954

**Comment by Mohamed3on** [comment]
Hi! is there any timeline for this feature?

**Comment by Tyriar** [comment]
Nope.

Here's the upstream issue https://github.com/sourcelair/xterm.js/issues/958

**Comment by coreh** [comment]
Would be really nice if there was a setting to re-enable the old terminal rendering until all the issues with the new terminal are resolved (blurriness, ligatures, emoji).

**Comment by MPV** [comment]
It seems there is now an `xterm.js` plugin for ligatures support:
https://github.com/xtermjs/xterm-addon-ligatures

How can we get it into vscode?

**Comment by Tyriar** [comment]
@MPV it would have to be built into the core, the main thing that needs to happen now is to evaluate how much weight the addon would add to the VS Code distributable.

**Comment by phavor** [comment]
![font-ligatures](https://user-images.githubusercontent.com/32136420/51835618-49667600-22fe-11e9-9e2b-25483c4970a0.png)


This is how my vscode term looks like, I want it to look like this:

![font-ligatures-2](https://user-images.githubusercontent.com/32136420/51832336-ab21e280-22f4-11e9-9083-51c449741be6.png)

**Comment by Tyriar** [comment]
@phavor I don't think your issue is about ligatures but rather your powerlines setup https://code.visualstudio.com/docs/editor/integrated-terminal#_can-i-use-powerline-fonts-in-the-integrated-terminal

**Comment by phavor** [comment]
@Tyriar I fixed though. Thanks

**Comment by FelikZ** [comment]
any news on it? I believe this is the last thing is missing for terminal to render everything proper

**Comment by wolf99** [comment]
Just tried Cascadia Code font in the integrated terminal... ligatures not present 😢

**Comment by jhpratt** [comment]
I see that there's a way to do this via an extension that modifies VS Code's internal files. Would it be possible to add a checkbox in settings that toggles this itself? @Tyriar

**Comment by Tyriar** [comment]
@jhpratt that's what it would do eventually, the remaining work here is slimming down the [ligatures xterm.js addon](https://github.com/xtermjs/xterm-addon-ligatures) to bare essentials and test/include it. I don't have time to do it right now due to other priorities though.

**Comment by Geobert** [comment]
Is this bug why I got this in vscode:
![image](https://user-images.githubusercontent.com/72570/67626136-2c37e700-f83f-11e9-899b-1f2b0331ec62.png)

instead of:
![image](https://user-images.githubusercontent.com/72570/67626144-35c14f00-f83f-11e9-9aa4-c85eb35b8b89.png)

(using this font https://github.com/adam7/delugia-code)?

**Comment by Tyriar** [comment]
@Geobert probably caused by https://github.com/xtermjs/xterm.js/issues/1709 or https://github.com/xtermjs/xterm.js/issues/942, ligatures are different to emoji.

**Comment by Tyriar** [comment]
Got a bunch of questions about ligatures when tweeting about the WebGL renderer so I'll give an update here. Here's the remaining work to do on this:

- xterm.js
  - Integrate [xtermjs/xterm-addon-ligatures](https://github.com/xtermjs/xterm-addon-ligatures) into the [xtermjs/xterm.js](https://github.com/xtermjs/xterm.js) repo, this will enable automated publishing and ensure tests continue to pass
  - Hook up ligatures into the demo
  - Set up some puppeteer tests that verify on all renderers (dom, canvas, webgl) that ligatures are working
  - Slim down the dependencies as much as possible
- vscode
  - Integrate xterm-addon-ligatures such that the addon code is not loaded if ligatures is off, add the setting, etc.

**Comment by threedogs1** [comment]
Apologies if this comment is not appropriately placed as I'm new to github and dev in general.

So it's my understanding that ligatures in the integrated terminal are disabled until other problems are fixed? Thanks and good day!

**Comment by Tyriar** [comment]
@threedogs1 yep, the first step above is happening in https://github.com/xtermjs/xterm.js/pull/2847

**Comment by regisbsb** [comment]
Looks like this has been merged? https://github.com/xtermjs/xterm.js/pull/2847

**Comment by regisbsb** [comment]
https://github.com/xtermjs/xterm.js/issues/958

**Comment by regisbsb** [comment]
This should help https://github.com/vercel/hyper/issues/3607

**Comment by Tyriar** [comment]
@regisbsb there are additional challenges for vscode, we would likely have to break the support once again soon as we're working towards turning off node in the renderer which the addon depends on.

**Comment by regisbsb** [comment]
@Tyriar I was really interest to see if it made to hyper tbh but vscode would be nice enough

**Comment by codemaster138** [comment]
Any progress on this? The Terminal still doesn't seem to support ligatures...

**Comment by Arelav** [comment]
> Any progress on this? The Terminal still doesn't seem to support ligatures...

This open source. You can always create PR if it's so urgent for you.

**Comment by Tyriar** [comment]
Here's what this issue is blocked on FYI https://github.com/xtermjs/xterm.js/issues/958#issuecomment-623114150

**Comment by Pyrolistical** [comment]
so this was released 2 months ago https://www.npmjs.com/package/xterm-addon-ligatures

does it just need integrated?

**Comment by Tyriar** [comment]
@Pyrolistical see https://github.com/microsoft/vscode/issues/34103#issuecomment-561851965

**Comment by zzstoatzz** [comment]
for anyone still looking who'd be satisfied with basic ligature support:

![image](https://user-images.githubusercontent.com/31014960/132166795-db105278-1fe0-4e33-94ab-145464703245.png)

using common zsh themes n fonts on big sur with the following entries in `settings.json`

```json5
{
    "terminal.integrated.fontFamily": "Fira Code", // 1st prompt, zsh 'cloud' theme

   // "terminal.integrated.fontFamily": "MesloLGS NF" // 2nd, oh-my-zsh p10k theme
}
```
`terminal.integrated.fontFamily` value set per discussion in this [gist](https://gist.github.com/480/3b41f449686a089f34edb45d00672f28)

doesn't seem to track output / totality of behaviour mentioned above, but for me those more a "nice-to-have"

**Comment by Mellbourn** [comment]
@zzstoatzz I'm sorry, but I don't quite understand your post. Where in the image are ligatures demonstrated?

**Comment by zzstoatzz** [comment]
> @zzstoatzz I'm sorry, but I don't quite understand your post. Where in the image are ligatures demonstrated?

oof, sincere apologies, I've deleted my comment since I realized I had misunderstood ligatures to be special characters as opposed to specific combinations of special characters

**Comment by Tyriar** [comment]
Blocker:

- [ ] Make sure font-finder isn't bundled into product build

**Comment by ToxicSmurf** [comment]
Around 2023 then?

**Comment by Tyriar** [comment]
I tried recently to integrate this but ran into some electron related problems including a segfault on macos, so this got pushed. See https://github.com/microsoft/vscode/pull/139537

**Comment by Tyriar** [comment]
PR to enable ligatures with an experimental setting: https://github.com/microsoft/vscode/pull/157008

**Comment by remidebette** [comment]
Hi,
So is this expected to be done?

**Comment by meganrogge** [comment]
@remidebette not yet, but soon

**Comment by JohnnyJTH** [comment]
So can I enable the experiment or is it not even an experiment yet?

**Comment by krischal111** [comment]
I want ligatures on my terminal, even if it's hacky. How hard would it be?

**Comment by bnb** [comment]
Would love to see this. Specifically, I'm using Twilio Sans Mono and would like the `(::)` ligature to work when I run `npx bitandbang` :)

**Comment by Tyriar** [comment]
This is still blocked on https://github.com/microsoft/vscode/pull/157008#issuecomment-1206848384. If anyone wants to help investigate that would be helpful, otherwise since this is pretty low priority against my other tasks it's up to whether I feel the motivation in my free time.

**Comment by Arelav** [comment]
> why is this feature taking several years what am I missing?

Yes, you are completely missing the fact it's an open source. Go and implement yourself if it's so urgent for you.

**Comment by junaga** [comment]
Claiming that VS Code is "open source" is childish and unreasonable. This is a Microsoft product.

**Comment by junaga** [comment]
[This is, open source](https://en.wikipedia.org/wiki/Copyleft), this ˅, isn't 
![image](https://github.com/microsoft/vscode/assets/35770913/72ff686f-3ae2-44fe-b516-a7ea153bd821)

I don't care but stop pretending the codebase has any excuse to be rotting, _Microsoft has enough cash and engineers_, it's just bad management, not "open source".

**Comment by Tyriar** [comment]
@junaga a few things:

- Copyleft is open source, open source is not copyleft
- This repository is [literally open source](https://github.com/microsoft/vscode/blob/main/LICENSE.txt)
- This particular issue would happen if someone helped out as called out in https://github.com/microsoft/vscode/issues/34103#issuecomment-1331018529, every time I give this a shot I hit some roadblock and it gets pushed back
- Suggesting to rewrite the terminal is really silly
- Suggesting to embed Windows Terminal is also silly, WT lacks a lot of features we do and it wouldn't work on the Linux, macOS or the web which is are all major platforms we target. You're either suggested we drop support for those platforms or waste time trying to support multiple terminal frontends
- [Read the code of conduct](https://opensource.microsoft.com/codeofconduct/)
- I recommend switching editors if ligatures is more important to you than everything else VS Code offers

**Comment by junaga** [comment]
### Windows Terminal

I was not aware Windows Terminal only runs on Windows, my apologies. `wt` being a new product, with the fancy trailer and all, I thought it was linux, macos and freebsd too. I still think it _a `wt` extension_ would have been worth a shot otherwise.

### open source

VS Code is, in any meaningful interpretation of the word "open source", not an open source community project, but a company-owned, brand, designed and tested. I know closed-source projects that are more open than the VS Code app. Publicly readable source code is of little use if everyone pulls your build with closed-source DRM and trackers. Also, there are no other providers to sign in and settings sync. My guess is ~80% of users don't know other builds exist, or they simply don't trust them (fear of scam). a gap in subject knowledge is the oldest excuse in the book of exploitation. The whole point of the term "open source" was freedom, for our half of the world, the blue internet. But I wouldn't expect an $MSFT employee to understand that.

again, I don't care,

### bad management

but hiding behind an "open source" shield is not respectable. you are employed and literally paid to write this code. I understand that ligatures are not as much a priority, as the fancy GitHub copilot chat, which is why I might just switch to Google IDX, they have ligatures on their homepage.

have a nice day @Tyriar

**Comment by Tyriar** [comment]
@junaga btw Google IDX is a fork of VS Code so it wouldn't have ligatures in the terminal either

**Comment by junaga** [comment]
god damn it, nvm then

**Comment by junaga** [comment]
but you see the issue, right? this is what MIT enables.

**Comment by Tyriar** [comment]
I'm not understanding your point. It's open source so people can fork it, which is a good thing? Google may also end up contributing upstream, improving VS Code and all its forks.

Yes VS Code is "built on OSS", not open source itself. It follows a very similar pattern to Chromium and Chrome. I'm pretty sure VS Code would not exist if there was not a private fork as MS needs to protect its trademarks. It seems like a net benefit to me that VS Code, this OSS project and all its forks exists.

You can argue forever about free software/copyleft vs open source, but that's not going to change the license of the repo or anything.

**Comment by junaga** [comment]
My point was, that it doesn't matter if a new fork is _better_ or not on MIT projects.

**Comment by junaga** [comment]
https://hackaday.com/2023/08/23/linux-fu-the-old-school-terminal/

**Comment by jeslinmx** [comment]
> just rewrite the terminal service If it's not feasible

> why is this feature taking several years what am I missing?

(https://github.com/microsoft/vscode/issues/34103#issuecomment-1680584127)

Look, I get that many of us have reasons to hate on Microsoft, but berating their engineers working on open-source (or at least, source-available) and free (as in beer) projects isn't a good way to cope with it, because:

1. They are engineers too, just like you (at least, I would like to believe you are, although the above quoted passages bring that into question)
2. It's free stuff, for goodness' sake. Even with paid products, companies sometimes fail to maintain adequately because their priorities are elsewhere, and the appropriate recourse in that case is to vote with your wallet and take your money elsewhere (or even invest in creating your own alternative out of spite). Open source isn't special, if you don't like it, take the money you're not paying elsewhere, make your own, or, even better, fix it yourself.

Once more for those in the back: open source isn't special. "submit your own patch" isn't a shield, it's literally the one unique thing you can do compared to closed-source, proprietary stuff.

On the other hand, demanding that some other engineer do it because they're paid (not by you) by the company funding the development of the project isn't open source, it's open entitlement. It's the type of thing that makes corporations feel like

> "open source" is childish and unreasonable

(https://github.com/microsoft/vscode/issues/34103#issuecomment-1684891222)

Thanks for coming to my TED talk, mods, please hide the above bunch of comments so other people who come looking for terminal ligature support won't go away shaking their hands and without terminal ligature support.

**Comment by junaga** [comment]
Hi @jeslinmx, That's a good TED talk. However, there are two things you got wrong.

## This is a Microsoft product

VS Code is a fork of gh:microsoft/vscode. However, the majority of users believe that VS Code is a distribution of gh:microsoft/vscode.

- gh:microsoft/vscode is MIT lincenced (not FOSS/copyleft)
- VS Code is closed-source and proprietary software
- VS Code includes evilish google grade, trackers, DRM, etc.
- the gh:microsoft/vscode repository description literally links to https://code.visualstudio.com/

This is a Microsoft product, with Microsoft responsibility, management, and funding.

The narrative of VS Code being open source was mistakenly constructed from misleading web pages, readmes, videos, and word-of-mouth conversations. Microsoft's failure to uphold its responsibility in this case is a reminder of the importance of corporate ethics.

Sorry @Tyriar, but comparing this distribution model to Google Chromium/Chrome is again a bit misleading. Google Chrome does not title itself as being "built on open Source" whatever that means. [Atom](https://atom.io/) was, but it had no users, which was a product design issue, not a licencing issue.

### Wikipedia got it right

https://en.wikipedia.org/wiki/Visual_Studio_Code

> made by Microsoft

https://en.wikipedia.org/wiki/Google_Chrome

> developed by Google

Not called "open source" once in either article.

## Source-available is the norm nowadays

Go ahead and try to get any downloads on a closed-source npm package. good luck. It's the same everywhere. The super secret SaaS product your company is working on would hardly get any downloads if distributed closed-source. Yes, we usually tend to overvalue our own software. But no developer will ever choose a closed-source technology (framework, library, middleware, code linter, etc.) to develop their own software with. VS Code and GitHub are an absolute exception to this, being the only closed-source non-infra technologies that the majority of developers use.

Oftentimes SaaS vendors don't even have the choice anymore to go closed-source. Without adoption, maybe even without VC investment, you can't build the product, and huge repercussions (https://www.youtube.com/watch?v=HzBA6FIn_Bo&t=142s) will follow if you close it afterward.

### Other software, imho

Only apps and games distributed within app stores benefit from being closed-source as of 2020. Users and developers expect products to be source-available. This is a significant change from 20 years ago. I would argue a positive cultural development. The world should not be split into computer wizards and mortals.

## Personal talk

How about you mark your own post as off-topic 😅. Because my previous comments were a lot more on-topic than your meta-discussion (discussion about discussions).

Of the many VS Code users I've spoken to, both in professional and personal settings, none were aware that VS Code is closed-source software. In fact, many Discord users I talked to (hundreds), were under the impression that it is free and open-source software (FOSS). It was important for me to speak out about this misunderstanding in plain and honest language.

Trying to personally attack ("you're not an engineer") is actually the thing that makes open source look unprofessional. **Not harsh language itself.** You're correct, I am not. I work as a developer, as you can read from my profile, and **I honestly don't have compassion for brittle, company, code**. and yes, grown-ups sometimes cuss. I apologize.

**Comment by Tyriar** [comment]
@junaga please stop, you're only being a nuisance. Whenever you make a comment, everyone who is subscribed or has commented will get a notification. A feature request about ligatures in a terminal is not the right place for this kind of thing.

You were warned before about the code of conduct. If you keep going I'll either block you from the repo or lock this issue, the latter of which I really don't want to do as it disables voting.

**Comment by junaga** [comment]
https://www.unicode.org/L2/L2023/23107-terminal-suppt.pdf TCSS (Terminal Complex Script Support)

> We find that a standardized specification for how to handle text in terminals is important work for the modern community of developers, and all other people that use text terminals.

> Fonts that meet these requirements, referred to as “terminal fonts”, will work properly with the layout algorithm defined in the specification.

Looks like the issue will be fixed upstream. Couldn't the mentioned algorithm be re-used, but with queries against the font instead? I mean, it would have to be implemented in the terminal emulator either way.

**Comment by ThaDaVos** [comment]
> @junaga the terminal renderers work by splitting up the rendering of the characters. The ligatures addon in xterm.js is implemented by joining sets of characters such that they are rendered together: https://github.com/xtermjs/xterm.js/tree/master/addons/xterm-addon-ligatures
> 
> So yes, but this is already done. We're just stuck on a packaging issue as mentioned in the last update: [#34103 (comment)](https://github.com/microsoft/vscode/issues/34103#issuecomment-1331018529)

The mentioned PR is closed but this issue is still open - any idea what the current state is?

**Comment by starball5** [comment]
Related on Stack Overflow:

- [setting character variants ss01, ss02, ... in vscode settings does not work in vscode terminal](https://stackoverflow.com/q/69379859/11107541)
- [Progress bar in Vs code terminal](https://stackoverflow.com/q/77730501/11107541)

**Comment by cocoonkid** [comment]
so this is still an  open  issue ? Is there a specific reason why ?

**Comment by airstrike** [comment]
Can we ever get this implemented? 2025 is around the corner and this issue was opened in 2017?

**Comment by bnb** [comment]
@airstrike if you want it, you should see if you can fix the previously mentioned blocker and submit a PR yourself: https://github.com/microsoft/vscode/issues/34103#issuecomment-1699403159

**Comment by darianmorat** [comment]
With this config: `"editor.fontLigatures": "'calt' 0, 'zero' 1"` in settings.json you can get the slashed zero for Jetbrains Mono, but terminal still shows the usual doted zero... A `"terminal.fontLigatures": "'calt' 0, 'zero' 1"` would be a great fit

**Comment by pattrickrice** [comment]
@Tyriar is there a repro of the bundling issue? I feel like that would be a good starting point for those that want to help here

**Comment by Tyriar** [comment]
Giving it another shot in https://github.com/microsoft/vscode/pull/232829

**Comment by pattrickrice** [comment]
Wow amazing!

**Comment by iwangbowen** [comment]
Finally

**Comment by joshlawton** [comment]
Just to confirm, this setting only accepts a boolean value and does not enable explicit font-feature-settings like `ss01` or `cv69`? That is, even if `terminal.integrated.fontLigatures` is true, the font may not necessarily match the style sets and character variants enabled in the editor?

```jsonc
// settings.json

// Works
"terminal.integrated.fontLigatures": true

// Won't work as expected at this time?
"terminal.integrated.fontLigatures": "'calt', 'liga', 'ss19', 'cv69'"
```

Compare [editorOptions.ts](https://github.com/microsoft/vscode/blob/d9c75e757228d1f272ab9a81a09171b37109e2e0/src/vs/editor/common/config/editorOptions.ts#L1789) to [terminalConfiguration.ts](https://github.com/microsoft/vscode/blob/d9c75e757228d1f272ab9a81a09171b37109e2e0/src/vs/workbench/contrib/terminal/common/terminalConfiguration.ts#L179).

**Comment by Tyriar** [comment]
@joshlawton yes it's all or nothing right now. If someone can figure out how to pass that information on to an OffscreenCanvas it should be trivial to add the support: https://developer.mozilla.org/en-US/docs/Web/API/OffscreenCanvasRenderingContext2D#text_styles

Created https://github.com/microsoft/vscode/issues/235316 to track

**Comment by AndreiKrutikov** [comment]
It works strangely on the latest Insiders build. Some of the ligatures are working, and some are not. Cannot figure out the rule, but for instance '=>' works, but not '-<'. 
Note that it doesn't work in the input but does work on the output; some ligatures don't work either on the input line or in the output.
![Image](https://github.com/user-attachments/assets/60ecfbca-e269-44c9-82a6-19131df9a5bc)

**Comment by Master-Hash** [comment]
> It works strangely on the latest Insiders build. Some of the ligatures are working, and some are not. Cannot figure out the rule, but for instance '=>' works, but not '-<'. 
> Note that it doesn't work in the input but does work on the output; some ligatures don't work either on the input line or in the output.
> ![Image](https://github.com/user-attachments/assets/60ecfbca-e269-44c9-82a6-19131df9a5bc)
>  

I found - and < has different colors in the first line, which may be the reason.

The third line is correct.

**Comment by Tyriar** [comment]
> I found - and < has different colors in the first line, which may be the reason.

@Master-Hash yes that's by design in that particular case since pwsh is coloring is differently

**Comment by zvdy** [comment]
# Inconsistent Ligatures 

This was just released as [preview](https://www.youtube.com/watch?v=faf6PhmgSzw) and for some reason is not properly doing the ligatures on zsh, tried a few `nerd fonts` too, I'm certain the font is not the issue as it works on the editor itself but not the terminal for some reason It's inconsistent and completely random, attaching image: 

![Image](https://github.com/user-attachments/assets/e2f05a48-930c-4cba-a52e-e375325f6647)

As you can see `<==` is properly represented but not the rest of ligatures also failing for the next <== which is typed in the same exact manner.
 
snippet settings.json

```js
{
  "terminal.integrated.fontFamily": "Fira Code",
  "editor.fontFamily": "Fira Code",

  "editor.fontLigatures": true,
  "terminal.integrated.fontWeight": "normal",
  "terminal.integrated.fontLigatures": true,
}
```

Any idea on why this behavior is happening only for certain ligatures, or any fixes/reconfiguration I can do ?

## System
- Arch Linux
- zsh
- P10k
- Fira Code

**Comment by sunshaoce** [comment]
I encountered a problem where the progress bar cannot be displayed correctly in the `terminal` using the `Fira Code` font.

```shell
echo "[=========================---------------------------------------------------------------]"
```
Will be
![Image](https://github.com/user-attachments/assets/e2d4f273-f475-4adb-9975-f97f561dd4ba)

**Comment by Tyriar** [comment]
@sunshaoce tracked in https://github.com/microsoft/vscode/issues/235988 👍
